File: cpp/bh_v4_threaded_tree.cpp
#include <iostream>
#include <vector>
#include "node.hpp"
#include <algorithm>
#include "timer.hpp"
#include <ctime>
#include <fstream> 
#include <cmath>
#include <iomanip>
#include "benchmark_options.hpp"

constexpr float G = 1.0f;
constexpr float TIME_STEP = 0.016f;
float THETA = 0.3f;
using Node = NodeV4;

struct Particle
{
    float velocityX{0.0f};
    float velocityY{0.0f};
    float accX{0.0f};
    float accY{0.0f};
    float mass{0.0f};
    float posX{0.0f};
    float posY{0.0f};
};

int getQuadrant(const Node& node, const Particle& particle) 
{
    int index = 0; 

    if (particle.posX > node.boundsX) 
    {
        index += 1; 
    }

    if (particle.posY > node.boundsY) 
    {
        index += 2; 
    }

    return index;
}
void insertParticle(int nodeIdx, int pIdx, std::vector<Node>& arena, std::vector<Particle>& particles)
{
    if (arena[nodeIdx].particleIndex != -1) 
    {
        int oldIdx = arena[nodeIdx].particleIndex;
        float shift = 0.0001f;
        while (particles[pIdx].posX == particles[oldIdx].posX && particles[pIdx].posY == particles[oldIdx].posY) 
        {
            particles[pIdx].posX += shift;
            shift *= 2.0f;
        }
    }

    if (arena[nodeIdx].particleIndex == -1 && arena[nodeIdx].children[0] == -1)
    {
        arena[nodeIdx].particleIndex = pIdx;
        return;
    }
    if (arena[nodeIdx].children[0] != -1)
    {
        int quadrant = getQuadrant(arena[nodeIdx], particles[pIdx]);
        insertParticle(arena[nodeIdx].children[quadrant], pIdx, arena, particles);
        return;
    }

    int oldPIdx = arena[nodeIdx].particleIndex;
    arena[nodeIdx].particleIndex = -1;
    for (int i = 0; i < 4; ++i)
    {
        Node child;
        child.halfSize = arena[nodeIdx].halfSize / 2.0f;
        child.boundsX = arena[nodeIdx].boundsX + ((i % 2) * 2 - 1) * child.halfSize;
        child.boundsY = arena[nodeIdx].boundsY + ((i / 2) * 2 - 1) * child.halfSize;
        arena[nodeIdx].children[i] = static_cast<int>(arena.size());
        arena.push_back(child);
    }
    insertParticle(nodeIdx, oldPIdx, arena, particles);
    insertParticle(nodeIdx, pIdx, arena, particles);
}

void computeMassDistribution(int nodeIdx, std::vector<Node>& arena, const std::vector<Particle>& particles)
{
    Node& node = arena[nodeIdx];
    if (node.children[0] != -1)
    {
        node.mass = 0.0f;
        node.centerX = 0.0f;
        node.centerY = 0.0f;
        for (int i = 0; i < 4; ++i)
        {
            int childIdx = node.children[i];
            computeMassDistribution(childIdx, arena, particles);
            float childMass = arena[childIdx].mass;
            node.mass += childMass;

            node.centerX += arena[childIdx].centerX * childMass;
            node.centerY += arena[childIdx].centerY * childMass;
        }
        if (node.mass > 0.0f)
        {
            node.centerX /= node.mass;
            node.centerY /= node.mass;
        }
    }

    else if(node.particleIndex != -1)
    {
        const Particle& p = particles[node.particleIndex];
        node.mass = p.mass;
        node.centerX = p.posX;
        node.centerY = p.posY;
    }
    else 
    {
        node.mass = 0.0f;
        node.centerX = 0.0f;
        node.centerY = 0.0f;
    }
}

void threadTree(int nodeIdx, int nextIdx, std::vector<Node>& arena)
{
    arena[nodeIdx].next = nextIdx;
    if (arena[nodeIdx].children[0] != -1)
    {
        for (int i = 0; i < 3; ++i)
        {
            threadTree(arena[nodeIdx].children[i], arena[nodeIdx].children[i + 1], arena);
        }
        threadTree(arena[nodeIdx].children[3], nextIdx, arena);
    }
}

void calculateForces(int pIdx, std::vector<Particle>& particles, const std::vector<Node>& arena)
{
    Particle& p = particles[pIdx];
    int currNodeIdx = 0; 

    while (currNodeIdx != -1)
    {
        const Node& node = arena[currNodeIdx];

        float dx = node.centerX - p.posX;
        float dy = node.centerY - p.posY;
        float distSq = dx * dx + dy * dy;

        if (distSq < 1e-5f) 
        {
            currNodeIdx = node.next;
            continue;
        }

        float side_length = node.halfSize * 2.0f;
        float r_c_sq = (side_length * side_length) * 0.5f; // (s_c * sqrt(2)/2)^2 = s_c^2 * 0.5
        if (r_c_sq < THETA * THETA * distSq || node.children[0] == -1)
        {
            float dist = std::sqrt(distSq);
            float acc = G * node.mass / (distSq + 1.0f);
            p.accX += acc * (dx / dist);
            p.accY += acc * (dy / dist);

            currNodeIdx = node.next;
        }
        else
        {
            currNodeIdx = node.children[0];
        }
    }
}

// struct PhysicsMetrics {
//     double totalMomentumX = 0.0;
//     double totalMomentumY = 0.0;
//     double totalKineticEnergy = 0.0;
//     double centerX = 0.0;
//     double centerY = 0.0;
// };

// PhysicsMetrics calculatePhysicsDiagnostics(const std::vector<Particle>& particles) {
//     PhysicsMetrics m;
//     double totalMass = 0.0;

//     for (const auto& p : particles) {
//         double mass = static_cast<double>(p.mass);
//         double vx = static_cast<double>(p.velocityX);
//         double vy = static_cast<double>(p.velocityY);
//         double px = static_cast<double>(p.posX);
//         double py = static_cast<double>(p.posY);

//         m.totalMomentumX += mass * vx;
//         m.totalMomentumY += mass * vy;
//         m.totalKineticEnergy += 0.5 * mass * (vx * vx + vy * vy);

//         m.centerX += mass * px;
//         m.centerY += mass * py;
//         totalMass += mass;
//     }

//     m.centerX /= totalMass;
//     m.centerY /= totalMass;

//     return m;
// }

void validateForceAccuracy(int currentFrame, const std::vector<Particle>& bh_particles)
{
    std::cout << "\n--- FORCE ACCURACY VALIDATION (Frame " << currentFrame << ") ---\n";

    double sum_diff_sq = 0.0;
    double sum_bf_sq = 0.0;
    std::vector<double> local_relative_errors;
    local_relative_errors.reserve(bh_particles.size());

    // Compute reference O(N^2) forces for the current Barnes-Hut positions
    //#pragma omp parallel for reduction(+:sum_diff_sq, sum_bf_sq) schedule(dynamic, 32)
    for (std::size_t i = 0; i < bh_particles.size(); ++i)
    {
        float exact_accX = 0.0f;
        float exact_accY = 0.0f;
        
        for (std::size_t j = 0; j < bh_particles.size(); ++j)
        {
            if (i == j) continue;
            float dx = bh_particles[j].posX - bh_particles[i].posX;
            float dy = bh_particles[j].posY - bh_particles[i].posY;
            float distSq = dx * dx + dy * dy;
            
            if (distSq < 1e-5f) continue;
            
            float dist = std::sqrt(distSq);
            float acc = G * bh_particles[j].mass / (distSq + 1.0f);
            exact_accX += acc * (dx / dist);
            exact_accY += acc * (dy / dist);
        }

        double diffX = bh_particles[i].accX - exact_accX;
        double diffY = bh_particles[i].accY - exact_accY;
        
        double diff_sq = diffX * diffX + diffY * diffY;
        double bf_sq = exact_accX * exact_accX + exact_accY * exact_accY;
        
        sum_diff_sq += diff_sq;
        sum_bf_sq += bf_sq;

        // Relative error for the 95th percentile
        double exact_norm = std::sqrt(bf_sq);
        double diff_norm = std::sqrt(diff_sq);
        if (exact_norm > 1e-6) {
            double rel_err = diff_norm / exact_norm;
            local_relative_errors.push_back(rel_err);
        }
    }

    double rms_error = std::sqrt(sum_diff_sq / sum_bf_sq);
    
    std::sort(local_relative_errors.begin(), local_relative_errors.end());
    double p95_error = 0.0;
    if (!local_relative_errors.empty()) {
        size_t p95_index = static_cast<size_t>(0.95 * local_relative_errors.size());
        p95_error = local_relative_errors[p95_index];
    }

    std::cout << std::fixed << std::setprecision(4);
    std::cout << "Global force error (RMS): " << rms_error * 100.0 << " %\n";
    std::cout << "95th percentile error:      " << p95_error * 100.0 << " %\n";
    std::cout << "--------------------------------------------------\n";
}

int mainMain(const BenchmarkOptions& options)
{
    srand(42);
    std::vector<Particle> particles;
    std::vector<Node> treeArena;
    //treeArena.reserve(particles.size() * 6);
    Timer timer;
    double totalTreeBuildTime = 0.0;
    double totalForceTime = 0.0;

   std::ifstream inFile(options.input);
    if (!inFile)
    {
        std::cerr << "Error: Could not open file " << options.input << "!\n";
        return 1;
    }

    Particle p;
    p.accX = 0.0f;
    p.accY = 0.0f;
    
    while (inFile >> p.posX >> p.posY >> p.velocityX >> p.velocityY >> p.mass)
    {
        particles.push_back(p);
    }
    inFile.close();
    if (!validateParticleCount(options, particles.size())) return 1;

    //treeArena.reserve(particles.size() * 6);

    for (int frame = 0; frame < options.totalFrames(); ++frame)
    {
        timer.start(); 
        float minX = particles[0].posX, maxX = particles[0].posX;
        float minY = particles[0].posY, maxY = particles[0].posY;
        
        for (const auto& particle : particles)
        {
            if (particle.posX < minX) minX = particle.posX;
            if (particle.posX > maxX) maxX = particle.posX;
            if (particle.posY < minY) minY = particle.posY;
            if (particle.posY > maxY) maxY = particle.posY;
        }

        float centerX = (minX + maxX) / 2.0f;
        float centerY = (minY + maxY) / 2.0f;
        float halfWidth = (maxX - minX) / 2.0f;
        float halfHeight = (maxY - minY) / 2.0f;
        float maxHalfSize = std::max(halfWidth, halfHeight) + 1.0f;

        treeArena.clear();
        Node root; 
        root.boundsX = centerX; 
        root.boundsY = centerY; 
        root.halfSize = maxHalfSize;
        treeArena.push_back(root);

        for (std::size_t i = 0; i < particles.size(); ++i)
        {
            insertParticle(0, static_cast<int>(i), treeArena, particles);
        }
        // if (frame == 0) {
        //     // Particle array size:
        //     size_t particlesMem = particles.capacity() * sizeof(Particle);
        //     // Maximum reserved tree-arena size:
        //     size_t treeMem = treeArena.capacity() * sizeof(Node);

        //     std::cout << std::fixed << std::setprecision(6);
        //     double totalAppMemMB = static_cast<double>(particlesMem + treeMem) / (1024.0 * 1024.0);
        //     std::cout << "Algorithm memory usage: " << totalAppMemMB << " MB\n";
        //     std::cout << "Created " << treeArena.size() << " tree nodes.\n";
        //     std::cout << "Size of Particle: " << sizeof(Particle) << " bytes\n";
        //     std::cout << "Size of Node (V4/V5): " << sizeof(Node) << " bytes\n";
        // }

        computeMassDistribution(0, treeArena, particles);
        threadTree(0, -1, treeArena);

        const double treeTime = timer.stopTime();
        if (frame >= options.warmupFrames) totalTreeBuildTime += treeTime;
        // if(frame == 0) {
        //     std::cout << "Tree construction time: " << (totalTreeBuildTime) << " ms\n";
        //     std::cout << "Tree construction cycles: " << std::fixed << (totalCyclesTree) << " cycles\n";
        // }

        timer.start();
        for (std::size_t i = 0; i < particles.size(); ++i)
        {
            calculateForces(static_cast<int>(i), particles, treeArena);
        }

        if (frame == options.warmupFrames && !writeForces(options.dumpForces, particles)) return 1;

        // if (frame == 0 or frame == FRAMES - 1 or frame == 150) { 
        //     validateForceAccuracy(frame, particles);
        // }
        // ---------------------------------------------------------

        // POSITION UPDATE AND FORCE RESET
        for (auto& particle : particles)
        {
            particle.velocityX += particle.accX * TIME_STEP;
            particle.velocityY += particle.accY * TIME_STEP;
            particle.posX += particle.velocityX * TIME_STEP;
            particle.posY += particle.velocityY * TIME_STEP;
            
            particle.accX = 0.0f; 
            particle.accY = 0.0f; 
        }
        const double forceTime = timer.stopTime();
        if (frame >= options.warmupFrames) totalForceTime += forceTime;
        // if (frame == 0 or frame == FRAMES - 1) {
        //     auto metrics = calculatePhysicsDiagnostics(particles);
        //     std::cout << "Frame " << frame << ":\n";
        //     std::cout << "Ped (" << metrics.totalMomentumX << ", " << metrics.totalMomentumY << ")\n";
        //     std::cout << "Energia kinetyczna " << metrics.totalKineticEnergy << "\n";
        //     std::cout << "Srodek masy (" << metrics.centerX << ", " << metrics.centerY << ")\n";
        // }
    }

    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Tree construction time: " << (totalTreeBuildTime / options.frames) << " ms / frame\n";
    std::cout << "Force/update time: " << (totalForceTime / options.frames) << " ms / frame\n";
    std::cout << "Cleanup time: 0.000000 ms / frame\n";
    std::cout << "Total measured time: " << (totalTreeBuildTime + totalForceTime) << " ms\n";
    // std::ifstream outFile("reference_5000k.txt");
    // if (!outFile) 
    // {
    //     std::cout << "file error.\n";
    // } 
    // else 
    // {
    //     float totalError = 0.0f;
    //     float maxError = 0.0f;
    //     float refX = 0.0f, refY = 0.0f;
        
    //     for (const auto& p : particles)
    //     {
    //         outFile >> refX >> refY;
            
    //         float dx = p.posX - refX;
    //         float dy = p.posY - refY;
    //         float currentError = std::sqrt(dx * dx + dy * dy);
    //         totalError += currentError;
    //         if (currentError > maxError) {
    //             maxError = currentError; 
    //         }
    //     }
    //     outFile.close();
        
    //     float meanAbsoluteError = totalError / NUM_PARTICLES;
    //     std::cout << "Mean absolute position error (MAE): " << meanAbsoluteError << " units\n";
    //     std::cout << "Maximum position error: " << maxError << " units\n";
    // }
    return 0;
}

int main(int argc, char** argv)
{
    BenchmarkOptions options;
    if (!parseBenchmarkOptions(argc, argv, options)) return argc > 1 ? 1 : 0;
    THETA = options.theta;
    return mainMain(options);
}

File: nbody_rust/src/bin/bh_v4_threaded_tree.rs
#![allow(non_snake_case)]

use std::time::Instant;
use std::sync::OnceLock;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

const TIME_STEP: f32 = 0.016;
const G : f32 = 1.0;
static THETA: OnceLock<f32> = OnceLock::new();

#[path = "../benchmark_options.rs"]
mod benchmark_options;

fn theta() -> f32 { *THETA.get().expect("benchmark options were not initialized") }

#[derive(Clone)]
struct Particle {
    velocity_x: f32,
    velocity_y: f32,
    acc_x: f32,
    acc_y: f32,
    pos_x: f32,
    pos_y: f32,
    mass: f32,
}

struct Node {
    bounds_x: f32,
    bounds_y: f32,
    half_size: f32,
    mass: f32,
    center_of_mass_x: f32,
    center_of_mass_y: f32,
    particle_index: usize,
    children: [usize; 4],
    next: usize,
}

impl Default for Node {
    fn default() -> Self {
        Node {
            bounds_x: 0.0,
            bounds_y: 0.0,
            half_size: 0.0,
            mass: 0.0,
            center_of_mass_x: 0.0,
            center_of_mass_y: 0.0,
            particle_index: usize::MAX,
            children: [usize::MAX, usize::MAX, usize::MAX, usize::MAX],
            next: usize::MAX,
        }
    }
}

fn getQuadrant(node: &Node, particle: &Particle) -> i32
{
    let mut quadrant = 0;
    if particle.pos_x > node.bounds_x { quadrant += 1; }
    if particle.pos_y > node.bounds_y { quadrant += 2; }
    quadrant
}

fn threadTree(nodeIdx: usize, nextIdx: usize, arena: &mut Vec<Node>)
{
    arena[nodeIdx].next = nextIdx;
    if arena[nodeIdx].children[0] != usize::MAX
    {
        for i in 0..3
        {
            threadTree(arena[nodeIdx].children[i], arena[nodeIdx].children[i + 1], arena);
        }
        threadTree(arena[nodeIdx].children[3], nextIdx, arena);
    }
}

fn insertParticle(nodeIdx: usize, pIdx: usize, arena: &mut Vec<Node>,  particles: &mut Vec<Particle>)
{
    //let node = &mut arena[nodeIdx];
    if arena[nodeIdx].particle_index != usize::MAX
    {
        let oldIdx: usize = arena[nodeIdx].particle_index;
        let mut shift = 0.0001;
        while particles[pIdx].pos_x == particles[oldIdx].pos_x && particles[pIdx].pos_y == particles[oldIdx].pos_y
        {
            particles[pIdx].pos_x += shift;
            shift *= 2.0;
        }
    }

    if arena[nodeIdx].particle_index == usize::MAX && arena[nodeIdx].children[0] == usize::MAX
    {
        arena[nodeIdx].particle_index = pIdx;
        return;
    }

    if arena[nodeIdx].children[0] != usize::MAX
    {
        let quad = getQuadrant(&arena[nodeIdx], &particles[pIdx]) as usize;
        insertParticle( arena[nodeIdx].children[quad], pIdx, arena, particles);
        return;
    }

    let oldPIdx = arena[nodeIdx].particle_index;
    arena[nodeIdx].particle_index = usize::MAX;
    let half_size = arena[nodeIdx].half_size / 2.0;
    for i in 0..4
    {
        let offset_x = ((i % 2) * 2) as f32 - 1.0;
        let offset_y = ((i / 2) * 2) as f32 - 1.0;
        arena[nodeIdx].children[i] = arena.len();

        arena.push(Node {
            bounds_x: arena[nodeIdx].bounds_x + offset_x * half_size,
            bounds_y: arena[nodeIdx].bounds_y + offset_y * half_size,
            half_size,
            mass: 0.0,
            center_of_mass_x: 0.0,
            center_of_mass_y: 0.0,
            particle_index: usize::MAX,
            children: [usize::MAX, usize::MAX, usize::MAX, usize::MAX],
            next: usize::MAX,
        });
    }
    insertParticle(nodeIdx, oldPIdx, arena, particles);
    insertParticle(nodeIdx, pIdx, arena, particles);
}

fn computeMassDistribution(nodeIdx: usize, arena: &mut Vec<Node>, particles: &Vec<Particle>)
{
    //let node = &mut arena[nodeIdx];
    if arena[nodeIdx].children[0] != usize::MAX
    {
        arena[nodeIdx].mass = 0.0;
        arena[nodeIdx].center_of_mass_x = 0.0;
        arena[nodeIdx].center_of_mass_y = 0.0;

        for i in 0..4
        {
            let childIdx = arena[nodeIdx].children[i];
            computeMassDistribution(childIdx, arena, particles);
            //let child = &arena[childIdx];
            arena[nodeIdx].mass += arena[childIdx].mass;
            arena[nodeIdx].center_of_mass_x += arena[childIdx].center_of_mass_x * arena[childIdx].mass;
            arena[nodeIdx].center_of_mass_y += arena[childIdx].center_of_mass_y * arena[childIdx].mass;
        }
        if arena[nodeIdx].mass > 0.0
        {
            arena[nodeIdx].center_of_mass_x /= arena[nodeIdx].mass;
            arena[nodeIdx].center_of_mass_y /= arena[nodeIdx].mass;
        }
    }
    else if arena[nodeIdx].particle_index != usize::MAX
    {
        let pIdx = arena[nodeIdx].particle_index;
        arena[nodeIdx].mass = particles[pIdx].mass;
        arena[nodeIdx].center_of_mass_x = particles[pIdx].pos_x;
        arena[nodeIdx].center_of_mass_y = particles[pIdx].pos_y;
    }
        else
    {
        arena[nodeIdx].mass = 0.0;
        arena[nodeIdx].center_of_mass_x = 0.0;
        arena[nodeIdx].center_of_mass_y = 0.0;
    }
}


fn calculateForces(pIdx: usize, arena: &Vec<Node>, particles: &mut Vec<Particle>)
{

    let mut currNodeIdx: usize = 0;
    while currNodeIdx != usize::MAX
    {
        let node = &arena[currNodeIdx];
        let dx = node.center_of_mass_x - particles[pIdx].pos_x;
        let dy = node.center_of_mass_y - particles[pIdx].pos_y;
        let dist_sq = dx * dx + dy * dy;
        //let dist = dist_sq.sqrt();

        if dist_sq < 1e-5 {
            currNodeIdx = node.next;
            continue;
        }

        let r_sq = 2.0 * node.half_size * node.half_size;
        if r_sq < theta() * theta() * dist_sq || node.children[0] == usize::MAX
        {
            let dist = dist_sq.sqrt();
            let acc = G * node.mass / (dist_sq + 1.0);
            particles[pIdx].acc_x += acc * (dx / dist);
            particles[pIdx].acc_y += acc * (dy / dist);
            currNodeIdx = node.next;
        }
        else
        {
            currNodeIdx = node.children[0];
        }
    }
}

#[allow(dead_code)]
fn validateForceAccuracy(current_frame: usize, bh_particles: &[Particle]) {
    println!("\n--- FORCE ACCURACY VALIDATION (Frame {}) ---", current_frame);

    let mut sum_diff_sq: f64 = 0.0;
    let mut sum_bf_sq: f64 = 0.0;
    let mut local_relative_errors = Vec::with_capacity(bh_particles.len());

    for i in 0..bh_particles.len() {
        let mut exact_acc_x: f32 = 0.0;
        let mut exact_acc_y: f32 = 0.0;

        for j in 0..bh_particles.len() {
            if i == j { continue; }
            let dx = bh_particles[j].pos_x - bh_particles[i].pos_x;
            let dy = bh_particles[j].pos_y - bh_particles[i].pos_y;
            let dist_sq = dx * dx + dy * dy;

            if dist_sq < 1e-5 { continue; }

            let dist = dist_sq.sqrt();
            let acc = G * bh_particles[j].mass / (dist_sq + 1.0);
            exact_acc_x += acc * (dx / dist);
            exact_acc_y += acc * (dy / dist);
        }

        let diff_x = (bh_particles[i].acc_x - exact_acc_x) as f64;
        let diff_y = (bh_particles[i].acc_y - exact_acc_y) as f64;

        let diff_sq = diff_x * diff_x + diff_y * diff_y;
        let bf_sq = (exact_acc_x * exact_acc_x + exact_acc_y * exact_acc_y) as f64;

        sum_diff_sq += diff_sq;
        sum_bf_sq += bf_sq;

        let exact_norm = bf_sq.sqrt();
        let diff_norm = diff_sq.sqrt();

        if exact_norm > 1e-6 {
            local_relative_errors.push(diff_norm / exact_norm);
        }
    }

    let rms_error = (sum_diff_sq / sum_bf_sq).sqrt();

    local_relative_errors.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let mut p95_error = 0.0;
    if !local_relative_errors.is_empty() {
        let p95_index = (0.95 * local_relative_errors.len() as f64) as usize;
        p95_error = local_relative_errors[p95_index];
    }

    println!("Global force error (RMS): {:.4} %", rms_error * 100.0);
    println!("95th percentile error:      {:.4} %", p95_error * 100.0);
    println!("--------------------------------------------------");
}

fn mainMain(options: &benchmark_options::BenchmarkOptions)
{
    let mut particles = Vec::new();
    let mut arena = Vec::new();

    let path = Path::new(&options.input);
    let file = match File::open(&path) {
        Ok(f) => f,
        Err(_) => {
            eprintln!("Error: Could not open input file {}.", options.input);
            std::process::exit(1);
        }
    };

    let reader = io::BufReader::new(file);

    for line in reader.lines() {
        let line = line.expect("Error while reading a line from the file");

        // Split the line by spaces
        let parts: Vec<&str> = line.split_whitespace().collect();

        if parts.len() == 5 {
            particles.push(Particle {
                pos_x: parts[0].parse().unwrap(),
                pos_y: parts[1].parse().unwrap(),
                velocity_x: parts[2].parse().unwrap(),
                velocity_y: parts[3].parse().unwrap(),
                mass: parts[4].parse().unwrap(),
                acc_x: 0.0,
                acc_y: 0.0,
            });
        }
    }

    let mut total_force_time_ms = 0.0;
    let mut total_tree_time_ms = 0.0;
    if let Err(message) = benchmark_options::validate_particle_count(options, particles.len()) {
        eprintln!("{message}");
        return;
    }

    for frame in 0..options.total_frames() {

        let mut start_time = Instant::now();

        let mut minX = particles[0].pos_x;
        let mut maxX = particles[0].pos_x;
        let mut minY = particles[0].pos_y;
        let mut maxY = particles[0].pos_y;

        for p in &particles
        {
            if p.pos_x < minX { minX = p.pos_x; }
            if p.pos_x > maxX { maxX = p.pos_x; }
            if p.pos_y < minY { minY = p.pos_y; }
            if p.pos_y > maxY { maxY = p.pos_y; }
        }

        let centerX = (minX + maxX) / 2.0;
        let centerY = (minY + maxY) / 2.0;
        let halfWidth = (maxX - minX) / 2.0;
        let halfHeight = (maxY - minY) / 2.0;
        let maxHalfSize = f32::max(halfWidth, halfHeight) + 1.0;

        let mut root: Node = Node::default();
        root.bounds_x = centerX;
        root.bounds_y = centerY;
        root.half_size = maxHalfSize;
        arena.clear();
        arena.push(root);

        for i in 0..particles.len()
        {
            insertParticle(0, i, &mut arena, &mut particles);
        }

        // if j == 0 {
        //     let particlesMem: usize = particles.capacity() * std::mem::size_of::<Particle>();
        //     let arenaMem: usize = arena.capacity() * std::mem::size_of::<Node>();
        //     let totalAppMemMB = (particlesMem + arenaMem) as f64 / (1024.0 * 1024.0);
        //     println!("Algorithm memory usage: {} MB", totalAppMemMB);
        //     println!("Created {} tree nodes", arena.len());
        //     println!("Size of Particle: {:.6} bytes", std::mem::size_of::<Particle>());
        //     println!("Size of Node (V4/V5): {:.6} bytes", std::mem::size_of::<Node>());
        // }

        computeMassDistribution(0, &mut arena, &particles);
        threadTree(0, usize::MAX, &mut arena);
        let tree_time_ms = start_time.elapsed().as_secs_f64() * 1000.0;
        if frame >= options.warmup_frames {
            total_tree_time_ms += tree_time_ms;
        }

        start_time = Instant::now();
        for i in 0..particles.len()
        {
            calculateForces(i, &mut arena, &mut particles);
        }

        if frame == options.warmup_frames {
            if let Some(path) = options.dump_forces.as_deref() {
                if let Err(message) = benchmark_options::write_forces(
                    path,
                    particles.iter().map(|particle| (particle.acc_x, particle.acc_y)),
                ) {
                    eprintln!("{message}");
                    return;
                }
            }
        }


        // if j == FRAMES - 1 || j == 0 || j == 150{
        //     validateForceAccuracy(j, &particles);
        // }

        for p in &mut particles
        {
            p.velocity_x += p.acc_x * TIME_STEP;
            p.velocity_y += p.acc_y * TIME_STEP;
            p.pos_x += p.velocity_x * TIME_STEP;
            p.pos_y += p.velocity_y * TIME_STEP;
            p.acc_x = 0.0;
            p.acc_y = 0.0;
        }
        let force_time_ms = start_time.elapsed().as_secs_f64() * 1000.0;
        if frame >= options.warmup_frames {
            total_force_time_ms += force_time_ms;
        }


        // if j == 0 || j == FRAMES - 1 {
        //     let metrics = calculate_physics_diagnostics(&particles);
        //     println!("Frame {}:", j);
        //     println!("Ped ({:.6}, {:.6})", metrics.total_momentum_x, metrics.total_momentum_y);
        //     println!("Energia kinetyczna {:.6}", metrics.total_kinetic_energy);
        //     println!("Srodek masy ({:.6}, {:.6})", metrics.center_x, metrics.center_y);
        // }
}

    println!("Tree construction time: {:.4} ms / frame", total_tree_time_ms / (options.frames as f64));
    println!("Force/update time: {:.4} ms / frame", total_force_time_ms / (options.frames as f64));
    println!("Cleanup time: 0.0000 ms / frame");
    println!("Total measured time: {:.4} ms", total_force_time_ms + total_tree_time_ms);

    // let ref_path = Path::new("reference_1000k.txt");
    // match File::open(&ref_path) {
    //     Ok(ref_file) => {
    //         let ref_reader = io::BufReader::new(ref_file);
    //         let mut total_error = 0.0f32;
    //         let mut particle_count = 0usize;
    //         let mut current_error = 0.0f32;
    //         let mut max_error = 0.0f32;

    //         for (idx, line) in ref_reader.lines().enumerate() {
    //             if idx >= NUM_PARTICLES {
    //                 break;
    //             }

    //             let line = match line {
    //                 Ok(l) => l,
    //                 Err(_) => break,
    //             };

    //             let parts: Vec<&str> = line.split_whitespace().collect();
    //             if parts.len() >= 2 {
    //                 let ref_x: f32 = match parts[0].parse() {
    //                     Ok(x) => x,
    //                     Err(_) => break,
    //                 };
    //                 let ref_y: f32 = match parts[1].parse() {
    //                     Ok(y) => y,
    //                     Err(_) => break,
    //                 };

    //                 let dx = particles[idx].pos_x - ref_x;
    //                 let dy = particles[idx].pos_y - ref_y;
    //                 current_error = (dx * dx + dy * dy).sqrt();
    //                 total_error += current_error;
    //                 if current_error > max_error {
    //                     max_error = current_error;
    //                 }
    //                 particle_count += 1;
    //             }
    //         }

    //         if particle_count > 0 {
    //             let mean_absolute_error = total_error / particle_count as f32;
    //             println!("Mean absolute position error (MAE): {:.6} units", mean_absolute_error);
    //             println!("Maximum position error: {:.6} units", max_error);
    //         }
    //     }
    //     Err(_) => {
    //         eprintln!("file error.");
    //     }
    // }
}

fn main() {
    let options = match benchmark_options::parse_options() {
        Ok(options) => options,
        Err(message) => { eprintln!("{message}"); std::process::exit(1); }
    };
    THETA.set(options.theta).expect("benchmark options initialized once");
    mainMain(&options);
}
